// @borsia/db/src/schema/index.ts
import {
  pgTable,
  uuid,
  varchar,
  text,
  timestamp,
  boolean,
  integer,
  decimal,
  jsonb,
  pgEnum,
  uniqueIndex,
  index,
} from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm";

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export const userRoleEnum = pgEnum("user_role", ["student", "admin"]);

export const scholarshipStatusEnum = pgEnum("scholarship_status", [
  "draft",
  "published",
  "archived",
]);

export const notificationTypeEnum = pgEnum("notification_type", [
  "match",
  "deadline",
  "application_update",
  "system",
]);

export const matchStatusEnum = pgEnum("match_status", [
  "pending",
  "viewed",
  "applied",
  "dismissed",
]);

// ---------------------------------------------------------------------------
// Users & authentication
// ---------------------------------------------------------------------------

export const users = pgTable(
  "users",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    email: varchar("email", { length: 255 }).notNull(),
    passwordHash: varchar("password_hash", { length: 255 }).notNull(),
    role: userRoleEnum("role").notNull().default("student"),
    isActive: boolean("is_active").notNull().default(true),
    emailVerifiedAt: timestamp("email_verified_at", { withTimezone: true }),
    lastLoginAt: timestamp("last_login_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdate(() => new Date()),
  },
  (table) => [uniqueIndex("users_email_idx").on(table.email)],
);

export const usersRelations = relations(users, ({ one, many }) => ({
  studentProfile: one(studentProfiles),
  savedScholarships: many(savedScholarships),
  notifications: many(notifications),
  scholarshipMatches: many(scholarshipMatches),
}));

// ---------------------------------------------------------------------------
// Student profiles
// ---------------------------------------------------------------------------

export const studentProfiles = pgTable(
  "student_profiles",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    firstName: varchar("first_name", { length: 100 }).notNull(),
    lastName: varchar("last_name", { length: 100 }).notNull(),
    dateOfBirth: timestamp("date_of_birth", { withTimezone: true }),
    phone: varchar("phone", { length: 30 }),
    country: varchar("country", { length: 100 }),
    city: varchar("city", { length: 100 }),
    educationLevel: varchar("education_level", { length: 100 }),
    fieldOfStudy: varchar("field_of_study", { length: 150 }),
    gpa: decimal("gpa", { precision: 4, scale: 2 }),
    ieltsScore: decimal("ielts_score", { precision: 3, scale: 1 }),
    toeflScore: integer("toefl_score"),
    graduationYear: integer("graduation_year"),
    bio: text("bio"),
    interests: jsonb("interests").$type<string[]>().default([]),
    skills: jsonb("skills").$type<string[]>().default([]),
    languages: jsonb("languages").$type<string[]>().default([]),
    preferences: jsonb("preferences")
      .$type<{
        minAmount?: number;
        maxAmount?: number;
        preferredCountries?: string[];
        preferredFields?: string[];
      }>()
      .default({}),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdate(() => new Date()),
  },
  (table) => [uniqueIndex("student_profiles_user_id_idx").on(table.userId)],
);

export const studentProfilesRelations = relations(
  studentProfiles,
  ({ one, many }) => ({
    user: one(users, {
      fields: [studentProfiles.userId],
      references: [users.id],
    }),
    savedScholarships: many(savedScholarships),
    scholarshipMatches: many(scholarshipMatches),
  }),
);

// ---------------------------------------------------------------------------
// Scholarships
// ---------------------------------------------------------------------------

export const scholarships = pgTable(
  "scholarships",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    title: varchar("title", { length: 255 }).notNull(),
    slug: varchar("slug", { length: 255 }).notNull(),
    description: text("description").notNull(),
    provider: varchar("provider", { length: 255 }).notNull(),
    amount: decimal("amount", { precision: 12, scale: 2 }),
    currency: varchar("currency", { length: 3 }).notNull().default("USD"),
    deadline: timestamp("deadline", { withTimezone: true }),
    applicationUrl: varchar("application_url", { length: 500 }),
    status: scholarshipStatusEnum("status").notNull().default("draft"),
    country: varchar("country", { length: 100 }),
    educationLevel: varchar("education_level", { length: 100 }),
    fieldOfStudy: varchar("field_of_study", { length: 150 }),
    minGpa: decimal("min_gpa", { precision: 4, scale: 2 }),
    minIelts: decimal("min_ielts", { precision: 3, scale: 1 }),
    minToefl: integer("min_toefl"),
    eligibilityCriteria: jsonb("eligibility_criteria")
      .$type<{
        countries?: string[];
        educationLevels?: string[];
        fieldsOfStudy?: string[];
        minAge?: number;
        maxAge?: number;
        requiredSkills?: string[];
        requiredLanguages?: string[];
        other?: string[];
      }>()
      .default({}),
    tags: jsonb("tags").$type<string[]>().default([]),
    featured: boolean("featured").notNull().default(false),
    createdById: uuid("created_by_id").references(() => users.id, {
      onDelete: "set null",
    }),
    publishedAt: timestamp("published_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdate(() => new Date()),
  },
  (table) => [
    uniqueIndex("scholarships_slug_idx").on(table.slug),
    index("scholarships_status_idx").on(table.status),
    index("scholarships_deadline_idx").on(table.deadline),
    index("scholarships_featured_idx").on(table.featured),
  ],
);

export const scholarshipsRelations = relations(scholarships, ({ one, many }) => ({
  createdBy: one(users, {
    fields: [scholarships.createdById],
    references: [users.id],
  }),
  savedBy: many(savedScholarships),
  matches: many(scholarshipMatches),
}));

// ---------------------------------------------------------------------------
// Saved scholarships
// ---------------------------------------------------------------------------

export const savedScholarships = pgTable(
  "saved_scholarships",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    scholarshipId: uuid("scholarship_id")
      .notNull()
      .references(() => scholarships.id, { onDelete: "cascade" }),
    notes: text("notes"),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    uniqueIndex("saved_scholarships_user_scholarship_idx").on(
      table.userId,
      table.scholarshipId,
    ),
    index("saved_scholarships_user_id_idx").on(table.userId),
  ],
);

export const savedScholarshipsRelations = relations(
  savedScholarships,
  ({ one }) => ({
    user: one(users, {
      fields: [savedScholarships.userId],
      references: [users.id],
    }),
    scholarship: one(scholarships, {
      fields: [savedScholarships.scholarshipId],
      references: [scholarships.id],
    }),
  }),
);

// ---------------------------------------------------------------------------
// Scholarship matching engine
// ---------------------------------------------------------------------------

export const scholarshipMatches = pgTable(
  "scholarship_matches",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    scholarshipId: uuid("scholarship_id")
      .notNull()
      .references(() => scholarships.id, { onDelete: "cascade" }),
    score: decimal("score", { precision: 5, scale: 2 }).notNull(),
    matchReasons: jsonb("match_reasons").$type<string[]>().default([]),
    status: matchStatusEnum("status").notNull().default("pending"),
    matchedAt: timestamp("matched_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    viewedAt: timestamp("viewed_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdate(() => new Date()),
  },
  (table) => [
    uniqueIndex("scholarship_matches_user_scholarship_idx").on(
      table.userId,
      table.scholarshipId,
    ),
    index("scholarship_matches_user_id_idx").on(table.userId),
    index("scholarship_matches_score_idx").on(table.score),
    index("scholarship_matches_status_idx").on(table.status),
  ],
);

export const scholarshipMatchesRelations = relations(
  scholarshipMatches,
  ({ one }) => ({
    user: one(users, {
      fields: [scholarshipMatches.userId],
      references: [users.id],
    }),
    scholarship: one(scholarships, {
      fields: [scholarshipMatches.scholarshipId],
      references: [scholarships.id],
    }),
  }),
);

// ---------------------------------------------------------------------------
// Notifications
// ---------------------------------------------------------------------------

export const notifications = pgTable(
  "notifications",
  {
    id: uuid("id").defaultRandom().primaryKey(),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    type: notificationTypeEnum("type").notNull(),
    title: varchar("title", { length: 255 }).notNull(),
    message: text("message").notNull(),
    data: jsonb("data")
      .$type<Record<string, unknown>>()
      .default({}),
    isRead: boolean("is_read").notNull().default(false),
    readAt: timestamp("read_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("notifications_user_id_idx").on(table.userId),
    index("notifications_is_read_idx").on(table.isRead),
    index("notifications_created_at_idx").on(table.createdAt),
  ],
);

export const notificationsRelations = relations(notifications, ({ one }) => ({
  user: one(users, {
    fields: [notifications.userId],
    references: [users.id],
  }),
}));

// ---------------------------------------------------------------------------
// Type exports
// ---------------------------------------------------------------------------

export type User = typeof users.$inferSelect;
export type NewUser = typeof users.$inferInsert;

export type StudentProfile = typeof studentProfiles.$inferSelect;
export type NewStudentProfile = typeof studentProfiles.$inferInsert;

export type Scholarship = typeof scholarships.$inferSelect;
export type NewScholarship = typeof scholarships.$inferInsert;

export type SavedScholarship = typeof savedScholarships.$inferSelect;
export type NewSavedScholarship = typeof savedScholarships.$inferInsert;

export type ScholarshipMatch = typeof scholarshipMatches.$inferSelect;
export type NewScholarshipMatch = typeof scholarshipMatches.$inferInsert;

export type Notification = typeof notifications.$inferSelect;
export type NewNotification = typeof notifications.$inferInsert;
