// @borsia/api-zod/src/schemas/index.ts
import { z } from "zod";

// ---------------------------------------------------------------------------
// Shared
// ---------------------------------------------------------------------------

export const paginationQuerySchema = z.object({
  page: z.coerce.number().int().min(1).default(1),
  limit: z.coerce.number().int().min(1).max(100).default(20),
});

export const idParamSchema = z.object({
  id: z.string().uuid(),
});

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

export const registerBodySchema = z.object({
  email: z.string().email(),
  password: z.string().min(8).max(128),
  firstName: z.string().min(1).max(100),
  lastName: z.string().min(1).max(100),
});

export const loginBodySchema = z.object({
  email: z.string().email(),
  password: z.string().min(1),
});

export const authResponseSchema = z.object({
  token: z.string(),
  user: z.object({
    id: z.string().uuid(),
    email: z.string().email(),
    role: z.enum(["student", "admin"]),
  }),
});

// ---------------------------------------------------------------------------
// Student profile
// ---------------------------------------------------------------------------

export const studentProfileSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  firstName: z.string(),
  lastName: z.string(),
  dateOfBirth: z.string().datetime().nullable(),
  phone: z.string().nullable(),
  country: z.string().nullable(),
  city: z.string().nullable(),
  educationLevel: z.string().nullable(),
  fieldOfStudy: z.string().nullable(),
  gpa: z.string().nullable(),
  ieltsScore: z.string().nullable(),
  toeflScore: z.number().int().nullable(),
  graduationYear: z.number().int().nullable(),
  bio: z.string().nullable(),
  interests: z.array(z.string()),
  skills: z.array(z.string()),
  languages: z.array(z.string()),
  preferences: z.object({
    minAmount: z.number().optional(),
    maxAmount: z.number().optional(),
    preferredCountries: z.array(z.string()).optional(),
    preferredFields: z.array(z.string()).optional(),
  }),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const meResponseSchema = z.object({
  id: z.string().uuid(),
  email: z.string().email(),
  role: z.enum(["student", "admin"]),
  isActive: z.boolean(),
  lastLoginAt: z.string().datetime().nullable(),
  createdAt: z.string().datetime(),
  profile: studentProfileSchema.nullable(),
});

export const updateStudentProfileBodySchema = z.object({
  firstName: z.string().min(1).max(100).optional(),
  lastName: z.string().min(1).max(100).optional(),
  dateOfBirth: z.string().datetime().optional(),
  phone: z.string().max(30).optional(),
  country: z.string().max(100).optional(),
  city: z.string().max(100).optional(),
  educationLevel: z.string().max(100).optional(),
  fieldOfStudy: z.string().max(150).optional(),
  gpa: z.coerce.number().min(0).max(4).optional(),
  ieltsScore: z.coerce.number().min(0).max(9).optional(),
  toeflScore: z.coerce.number().int().min(0).max(120).optional(),
  graduationYear: z.coerce.number().int().optional(),
  bio: z.string().max(2000).optional(),
  interests: z.array(z.string()).optional(),
  skills: z.array(z.string()).optional(),
  languages: z.array(z.string()).optional(),
  preferences: z
    .object({
      minAmount: z.number().optional(),
      maxAmount: z.number().optional(),
      preferredCountries: z.array(z.string()).optional(),
      preferredFields: z.array(z.string()).optional(),
    })
    .optional(),
});

// ---------------------------------------------------------------------------
// Scholarships
// ---------------------------------------------------------------------------

export const scholarshipStatusSchema = z.enum(["draft", "published", "archived"]);

export const scholarshipSchema = z.object({
  id: z.string().uuid(),
  title: z.string(),
  slug: z.string(),
  description: z.string(),
  provider: z.string(),
  amount: z.string().nullable(),
  currency: z.string(),
  deadline: z.string().datetime().nullable(),
  applicationUrl: z.string().nullable(),
  status: scholarshipStatusSchema,
  country: z.string().nullable(),
  educationLevel: z.string().nullable(),
  fieldOfStudy: z.string().nullable(),
  minGpa: z.string().nullable(),
  minIelts: z.string().nullable(),
  minToefl: z.number().int().nullable(),
  eligibilityCriteria: z.object({
    countries: z.array(z.string()).optional(),
    educationLevels: z.array(z.string()).optional(),
    fieldsOfStudy: z.array(z.string()).optional(),
    minAge: z.number().optional(),
    maxAge: z.number().optional(),
    requiredSkills: z.array(z.string()).optional(),
    requiredLanguages: z.array(z.string()).optional(),
    other: z.array(z.string()).optional(),
  }),
  tags: z.array(z.string()),
  featured: z.boolean(),
  createdById: z.string().uuid().nullable(),
  publishedAt: z.string().datetime().nullable(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const scholarshipListQuerySchema = paginationQuerySchema.extend({
  status: scholarshipStatusSchema.optional(),
  country: z.string().optional(),
  fieldOfStudy: z.string().optional(),
  educationLevel: z.string().optional(),
  search: z.string().optional(),
  featured: z.coerce.boolean().optional(),
});

export const createScholarshipBodySchema = z.object({
  title: z.string().min(1).max(255),
  slug: z.string().min(1).max(255),
  description: z.string().min(1),
  provider: z.string().min(1).max(255),
  amount: z.coerce.number().positive().optional(),
  currency: z.string().length(3).default("USD"),
  deadline: z.string().datetime().optional(),
  applicationUrl: z.string().url().optional(),
  status: scholarshipStatusSchema.default("draft"),
  country: z.string().max(100).optional(),
  educationLevel: z.string().max(100).optional(),
  fieldOfStudy: z.string().max(150).optional(),
  minGpa: z.coerce.number().min(0).max(4).optional(),
  minIelts: z.coerce.number().min(0).max(9).optional(),
  minToefl: z.coerce.number().int().min(0).max(120).optional(),
  eligibilityCriteria: z
    .object({
      countries: z.array(z.string()).optional(),
      educationLevels: z.array(z.string()).optional(),
      fieldsOfStudy: z.array(z.string()).optional(),
      minAge: z.number().optional(),
      maxAge: z.number().optional(),
      requiredSkills: z.array(z.string()).optional(),
      requiredLanguages: z.array(z.string()).optional(),
      other: z.array(z.string()).optional(),
    })
    .optional(),
  tags: z.array(z.string()).optional(),
  featured: z.boolean().optional(),
});

export const updateScholarshipBodySchema = createScholarshipBodySchema.partial();

// ---------------------------------------------------------------------------
// Saved scholarships
// ---------------------------------------------------------------------------

export const savedScholarshipSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  scholarshipId: z.string().uuid(),
  notes: z.string().nullable(),
  createdAt: z.string().datetime(),
  scholarship: scholarshipSchema.optional(),
});

export const saveScholarshipBodySchema = z.object({
  scholarshipId: z.string().uuid(),
  notes: z.string().max(1000).optional(),
});

// ---------------------------------------------------------------------------
// Scholarship matches
// ---------------------------------------------------------------------------

export const matchStatusSchema = z.enum(["pending", "viewed", "applied", "dismissed"]);

export const scholarshipMatchSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  scholarshipId: z.string().uuid(),
  score: z.string(),
  matchReasons: z.array(z.string()),
  status: matchStatusSchema,
  matchedAt: z.string().datetime(),
  viewedAt: z.string().datetime().nullable(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  scholarship: scholarshipSchema.optional(),
});

export const matchListQuerySchema = paginationQuerySchema.extend({
  status: matchStatusSchema.optional(),
  minScore: z.coerce.number().min(0).max(100).optional(),
});

export const updateMatchStatusBodySchema = z.object({
  status: matchStatusSchema,
});

// ---------------------------------------------------------------------------
// Notifications
// ---------------------------------------------------------------------------

export const notificationTypeSchema = z.enum([
  "match",
  "deadline",
  "application_update",
  "system",
]);

export const notificationSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  type: notificationTypeSchema,
  title: z.string(),
  message: z.string(),
  data: z.record(z.unknown()),
  isRead: z.boolean(),
  readAt: z.string().datetime().nullable(),
  createdAt: z.string().datetime(),
});

export const notificationListQuerySchema = paginationQuerySchema.extend({
  isRead: z.coerce.boolean().optional(),
  type: notificationTypeSchema.optional(),
});

// ---------------------------------------------------------------------------
// Admin
// ---------------------------------------------------------------------------

export const adminStatsSchema = z.object({
  totalUsers: z.number().int(),
  totalStudents: z.number().int(),
  totalScholarships: z.number().int(),
  publishedScholarships: z.number().int(),
  totalMatches: z.number().int(),
  totalNotifications: z.number().int(),
});

// ---------------------------------------------------------------------------
// API response wrappers
// ---------------------------------------------------------------------------

export const paginatedResponseSchema = <T extends z.ZodTypeAny>(itemSchema: T) =>
  z.object({
    items: z.array(itemSchema),
    total: z.number().int(),
    page: z.number().int(),
    limit: z.number().int(),
    totalPages: z.number().int(),
  });

export const errorResponseSchema = z.object({
  error: z.string(),
  details: z.unknown().optional(),
});

// ---------------------------------------------------------------------------
// Type exports
// ---------------------------------------------------------------------------

export type RegisterBody = z.infer<typeof registerBodySchema>;
export type LoginBody = z.infer<typeof loginBodySchema>;
export type AuthResponse = z.infer<typeof authResponseSchema>;
export type MeResponse = z.infer<typeof meResponseSchema>;
export type StudentProfile = z.infer<typeof studentProfileSchema>;
export type UpdateStudentProfileBody = z.infer<typeof updateStudentProfileBodySchema>;
export type Scholarship = z.infer<typeof scholarshipSchema>;
export type ScholarshipListQuery = z.infer<typeof scholarshipListQuerySchema>;
export type CreateScholarshipBody = z.infer<typeof createScholarshipBodySchema>;
export type UpdateScholarshipBody = z.infer<typeof updateScholarshipBodySchema>;
export type SavedScholarship = z.infer<typeof savedScholarshipSchema>;
export type SaveScholarshipBody = z.infer<typeof saveScholarshipBodySchema>;
export type ScholarshipMatch = z.infer<typeof scholarshipMatchSchema>;
export type MatchListQuery = z.infer<typeof matchListQuerySchema>;
export type UpdateMatchStatusBody = z.infer<typeof updateMatchStatusBodySchema>;
export type Notification = z.infer<typeof notificationSchema>;
export type NotificationListQuery = z.infer<typeof notificationListQuerySchema>;
export type AdminStats = z.infer<typeof adminStatsSchema>;
